internal/eventstore/internal/repository/sql/push.go

```diff
@@ -3,26 +3,22 @@ package sql
 import (
 	"context"
 	"database/sql"
+	"errors"
 
 	"github.com/caos/logging"
-	"github.com/caos/zitadel/internal/errors"
+	caos_errs "github.com/caos/zitadel/internal/errors"
 	"github.com/caos/zitadel/internal/eventstore/models"
 	"github.com/cockroachdb/cockroach-go/crdb"
 )
 
-const insertStmt = "insert into eventstore.events " +
+const insertStmt = "INSERT INTO eventstore.events " +
 	"(event_type, aggregate_type, aggregate_id, aggregate_version, creation_date, event_data, editor_user, editor_service, resource_owner, previous_sequence) " +
-	"select $1, $2, $3, $4, coalesce($5, now()), $6, $7, $8, $9, " +
-	// case is to set the highest sequence or NULL in previous_sequence
-	"case (select exists(select event_sequence from eventstore.events where aggregate_type = $10 AND aggregate_id = $11)) " +
-	"WHEN true then (select event_sequence from eventstore.events where aggregate_type = $12 AND aggregate_id = $13 order by event_sequence desc limit 1) " +
-	"ELSE NULL " +
-	"end " +
-	"where (" +
-	// exactly one event of requested aggregate must have a >= sequence (last inserted event)
-	"(select count(id) from eventstore.events where event_sequence >= COALESCE($14, 0) AND aggregate_type = $15 AND aggregate_id = $16) = 1 OR " +
-	// previous sequence = 0, no events must exist for the requested aggregate
-	"((select count(id) from eventstore.events where aggregate_type = $17 and aggregate_id = $18) = 0 AND COALESCE($19, 0) = 0)) " +
+	"SELECT $1, $2, $3, $4, COALESCE($5, now()), $6, $7, $8, $9, $10 " +
+	"WHERE EXISTS (SELECT 1 WHERE " +
+	// exactly one event of requested aggregate must have the given previous sequence as sequence (last inserted event)
+	"EXISTS (SELECT 1 FROM eventstore.events WHERE event_sequence = COALESCE($11, 0) AND aggregate_type = $12 AND aggregate_id = $13) OR " +
```

Comment:
Doesn't this line still cause the problem to exist?
What about making the sequence 1-based and the previousSequence 0-based? I think that would make things much easier and would also be easier to understand.
You could programmatically allow event sequences > 0 only

Reply:
the `previous_sequence`-field must be unique (unique-constraint on table definition). That's why it must allow `NULL`. If we would use 0 instead of `NULL` we had to implement the constraint by ourselves, thats why `NULL` is allowed.
the `event_sequence`-field starts by one, or the value defined in the `event_sequence`-sequence. As you can see you cannot define the `event_sequence` programatically. It's always set by the database.

Reply:
The line doesn't cause the problem anymore because the `previous_sequence`-field is not calculated anymore by the database.